csharp/src/Ice/Connection.cs

```diff
@@ -331,50 +331,50 @@ public async ValueTask HeartbeatAsync(IProgress<bool>? progress = null, Cancella
             progress?.Report(true);
         }
 
-        /// <summary>Sets a close callback on the connection. The callback is called by the connection when it's
-        /// closed. If the callback needs more information about the closure, it can call Connection.throwException.
-        /// </summary>
-        /// <param name="callback">The close callback object.</param>
-        public void SetCloseCallback(Action<Connection> callback)
+        /// <summary>This event is raised when the connection is closed. If the connection is already closed when
+        /// subscribing the event is raised immediately. If the subscriber needs more information about the closure, it
+        /// can call Connection.ThrowException.</summary>
+        public event EventHandler? Closed
         {
-            lock (_mutex)
+            add
             {
-                if (_state >= ConnectionState.Closed)
+                lock (_mutex)
                 {
-                    if (callback != null)
+                    if (_state >= ConnectionState.Closed)
                     {
-                        Task.Run(() =>
-                        {
-                            try
-                            {
-                                callback(this);
-                            }
-                            catch (Exception ex)
-                            {
-                                _communicator.Logger.Error($"connection callback exception:\n{ex}\n{this}");
-                            }
-                        });
+                        Task.Run(() => value?.Invoke(this, EventArgs.Empty));
                     }
+                    _closed += value;
                 }
-                else
+            }
+            remove
+            {
+                lock (_mutex)
                 {
-                    _closeCallback = callback;
+                    _closed -= value;
                 }
             }
         }
 
-        /// <summary>Sets a heartbeat callback on the connection. The callback is called by the connection when a
-        /// heartbeat is received.</summary>
-        /// <param name="callback">The heartbeat callback object.</param>
-        public void SetHeartbeatCallback(Action<Connection> callback)
+        /// <summary>This event is raised when the connection receives a heartbeat.</summary>
+        public event EventHandler? HeartbeatReceived
```

Comment:
I would just do a simple:
```public event EventHandler? HeartbeatReceived;```

Why worry about handlers added after the connection is closed?
And I believe adding/removing from an event is thread-safe.

Reply:
This was just to keep the behavior of reporting the event if you subscribe to a closed connection, but if we don't care about this, it would be fine to simplify this and use the default add/remove.

Reply:
For heartbeats, we don't care.
For closed connection, we care and we should either execute the new handler (as you did) or throw an exception (as Benoit suggested).

Reply:
any preferences?

Reply:
Yes, we can execute the handler, that's fine. And, as you previously did, you'll need to keep the lock while adding the new handler to _closed to avoid a race condition.